File: c#/2020/Day08-2020.cs
using System;
using System.IO;
using System.Collections.Generic;
using System.Linq;

namespace AdventOfCode.Year2020
{
	public class Day08
	{
		List<Instruction> instructions = new List<Instruction>();

		public Day08()
		{
			var input = File.ReadAllLines("2020/day08.data");

			foreach (var line in input)
			{
				instructions.Add(new Instruction(line));
			}

			Part1();
			Part2();
		}

		void Part1()
		{
			var computer = new Computer();
			computer.Load(instructions);

			computer.Run();

			Console.WriteLine(computer.Accumulator);
		}

		void Part2()
		{
			var computer = new Computer();
			computer.Load(instructions);

			var error = false;

			int lastInstructionIndex = 0;

			do
			{
				error = computer.Run();

				// Mutate nop to jmp or jmp to nop
				if (error)
				{
					var newInstructions = new List<Instruction>(instructions);

					newInstructions.Find((i) => i.Operation == "nop" || i.Operation == "jmp");
					var index = newInstructions.FindIndex(lastInstructionIndex + 1, (i) => i.Operation == "nop" || i.Operation == "jmp");

					var instruction = newInstructions[index];
					instruction.Operation = instruction.Operation == "nop" ? "jmp" : "nop";
					newInstructions[index] = instruction;

					lastInstructionIndex = index;

					computer.Load(newInstructions);
				}
			} while (error);
			

			Console.WriteLine(computer.Accumulator);
		}
	}

	struct Instruction
	{
		public string Operation;
		public int Argument;

		public Instruction(string line)
		{
			var parts = line.Split(" ");

			Operation = parts[0];
			Argument = int.Parse(parts[1]);
		}
	}

	class Computer
	{
		public int PC = 0;
		public int Accumulator = 0;

		public List<Instruction> Instructions;

		public void Load(List<Instruction> instructions)
		{
			Instructions = instructions;

			Reset();
		}

		void Reset()
		{
			PC = 0;
			Accumulator = 0;
		}

		public bool Run()
		{
			var addresses = new HashSet<int>();

			while (PC < Instructions.Count)
			{
				addresses.Add(PC);

				var instruction = Instructions[PC];

				switch (instruction.Operation)
				{
					case "acc":
						Accumulator += instruction.Argument;
						PC++;
						break;

					case "nop":
						PC++;
						break;

					case "jmp":
						PC += instruction.Argument;
						break;
				}

				if (addresses.Contains(PC))
				{
					return true;
				}

			}

			return false;
		}
	}
}

File: swift/2020/Day08-2020.swift
import Foundation


fileprivate var program = [Instruction]()

func Day08_2020() {
    let input = try! String(contentsOfFile: "2020/day08.data")
    program = input.split(separator: "\n").map { Instruction(line: String($0)) }
    
    Part1()
    Part2()
}

fileprivate func Part1() {
    
    let computer = Computer()
    computer.Load(instructions: program)
    _ = computer.Run()
    
    print(computer.accumulator)
   
}

fileprivate func Part2() {
    let computer = Computer()
    computer.Load(instructions: program)
    
    var error = true
    
    var lastChange = 0
    
    var newInstructions = program
    
    repeat {
        computer.Load(instructions: newInstructions)
        computer.Reset()
        error = computer.Run()
        
        if error {
            newInstructions = program
            
            // change a nop to jmp or jmp to nop
            if let address = findFrom(instructions: newInstructions, from: lastChange + 1) {
                lastChange = address
                var newInstruction = newInstructions[address]
                newInstruction.operation = newInstruction.operation == "nop" ? "jmp" : "nop"
                newInstructions[address] = newInstruction
            }
        }
    } while error
    
    print(computer.accumulator)
}

func findFrom(instructions: [Instruction], from: Int) -> Int? {
    for i in from..<instructions.count {
        if instructions[i].operation == "nop" ||  instructions[i].operation == "jmp"{
            return i
        }
    }
    
    return nil
}

struct Instruction {
    var operation: String
    var argument: Int
    
    init(line: String) {
        let x = line.split(separator: " ")
        
        operation = String(x[0])
        argument = Int(x[1])!
    }
}

class Computer {
    
    var pc = 0
    var accumulator = 0
    
    var instructions = [Instruction]()
    
    func Load(instructions: [Instruction]) {
        self.instructions = instructions
        
        Reset()
    }
    
    func Reset() {
        pc = 0
        accumulator = 0
    }
    
    func Run() -> Bool {
        var executedAddresses = Set<Int>()
        var error = false
        
        while !error && pc < instructions.count {
            
            executedAddresses.insert(pc)
            
            let instruction = instructions[pc]
            
            switch instruction.operation {
            case "acc":
                accumulator += instruction.argument
                pc += 1

            case "nop":
                pc += 1

            case "jmp":
                pc += instruction.argument
            
            default:
                error = true
                print("invalid")
            }
            
            if executedAddresses.contains(pc) {
                error = true
            }
        }

        return error
    }
}
